### PR TITLE
New bitcoin_client in rust

### PR DESCRIPTION
**Pull request by edouardparis**
This PR adds a new `bitcoin_client_rs` directory.
The crate does not support the communication layer between the device 
and the client but the messaging protocol between them. 

TODO:

- [ ] Add more test cases
- [ ] fix comments and documentation

---

# Ledger Bitcoin application client

Client library in [Rust](https://www.rust-lang.org/) for the Ledger 
Bitcoin application with minimal dependencies.

If you wish to contribute to this library, please read
[CONTRIBUTING.md](CONTRIBUTING.md).

## Minimum Supported Rust Version

`bitcoin_client_rs` should always compile using **Rust 1.43** and pass
tests using **Rust 1.49**.

## Getting started

The `client::BitcoinClient` struct implements the methods that call and
interpret the commands between the Ledger device and your software. 

```rust
pub struct BitcoinClient<T: Transport> {...}
impl<T: Transport> BitcoinClient<T> {
    pub fn get_extended_pubkey(
        &self,
        path: &bitcoin::util::bip32::DerivationPath,
        display: bool,
    ) -> Result<bitcoin::util::bip32::ExtendedPubKey, BitcoinClientError<T::Error>>;
}
```

It requires an internal connection implementing the `client::Transport`
Trait.

```rust
pub trait Transport {
    type Error: Debug;
    fn exchange(&self, command: &APDUCommand) -> Result<(StatusWord, Vec<u8>), Self::Error>;
}
```

In order to satisfy this Trait, it is possible to import the
`ledger-transport-hid` crate from https://github.com/Zondax/ledger-rs.
Please, read the `examples/ledger_hwi/src/transport.rs` file to find an example.

## The `async` feature

The optional feature `async` adds the `async_client` module to the crate
and imports the `async_trait` library. The `async_client::BitcoinClient` 
struct is an asynchronous equivalent to the `BitcoinClient` struct. It 
requires an internal connection implementing the `async_client::Transport` Trait.

```rust
#[async_trait]
pub trait Transport {
    type Error: Debug;
    async fn exchange(&self, command: &APDUCommand) -> Result<(StatusWord, Vec<u8>), Self::Error>;
}
```

## The `no-std` support

Work in progress.

## Example

The code source for a simple tool to communicate with either a Ledger device or Speculos
emulator can be found in the `examples` directory.

Example of a command to retrieve the extended pubkey with the given
derivation path and display it on the device screen:  
```
cargo run --package ledger_hwi -- \
get-extended-pubkey --derivation-path "m/44'/0'/0'/0/0" --display
```

### CODE REVIEW

**Comment by darosior** [comment]
Any reason for the choice of MSRV?

**Review by bigspider — APPROVED** [review]
Added a few nits, but happy to merge even without them, and improve from there.
Thanks a lot @edouardparis, this is going to be very useful!

**Comment by bigspider** [comment]
Added rust tests to the CI (ignore CodeQL and Sonarcloud, they're broken on our end).

File: bitcoin_client_rs/src/common.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            core::u32::MAX as usize
+        );
+        assert_eq!(
+            read_varint(&write_varint(core::u64::MAX as usize))
+                .unwrap()
+                .0,
+            core::u64::MAX as usize
+        );
+    }
+}
```
[the remark this answers is not in the document]
Comment (edouardparis) [reply bitcoin_client_rs/src/common.rs]: I removed read_varint and write_varint for https://docs.rs/bitcoin/latest/bitcoin/consensus/encode/struct.VarInt.html

File: bitcoin_client_rs/src/wallet.rs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+    hashes::{sha256, Hash, HashEngine},
+    util::bip32::{DerivationPath, Error, ExtendedPubKey, Fingerprint, KeySource},
+};
+
+use crate::{common::write_varint, merkle::MerkleTree};
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Version {
+    V1 = 1,
+    V2 = 2,
```
[the remark this answers is not in the document]
Comment (edouardparis) [reply bitcoin_client_rs/src/wallet.rs:16]: I would prefer to not touch anymore the module if you do not mind. The support is done and tested and maybe the whole wallet module will be refactored in a next version to support an optional `miniscript` feature to make use of [DescriptorXKey struct](https://docs.rs/miniscript/latest/miniscript/descriptor/struct.DescriptorXKey.html)

**Pull request merged**